util/clusterauth/clusterauth.go

```diff
@@ -215,39 +216,103 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// createServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func createServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.Poll(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+		if getErr != nil {
+			return false, getErr
+		}
+		if len(serviceAccount.Secrets) == 0 {
+			return false, nil
 		}
-		// Scan all secrets looking for one of the correct type:
-		for _, oRef := range serviceAccount.Secrets {
-			var getErr error
-			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
+		return true, nil
+	})
+	if errors.IsNotFound(err) {
+		return "", err
+	}
+
+	if len(serviceAccount.Secrets) != 0 {
+		for _, s := range serviceAccount.Secrets {
+			existingSecret, err := clientset.CoreV1().Secrets(ns).Get(context.Background(), s.Name, metav1.GetOptions{})
 			if err != nil {
-				return false, fmt.Errorf("Failed to retrieve secret %q: %v", oRef.Name, getErr)
+				return "", fmt.Errorf("Failed to retrieve secret %q: %v", s.Name, err)
 			}
-			if secret.Type == corev1.SecretTypeServiceAccountToken {
-				return true, nil
+
+			if existingSecret.Type == corev1.SecretTypeServiceAccountToken {
+				return existingSecret.Name, nil
 			}
 		}
-		return false, nil
-	})
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: serviceAccount.Name + "-token-",
+			Namespace:    serviceAccount.Namespace,
+			Annotations: map[string]string{
+				corev1.ServiceAccountNameKey: serviceAccount.Name,
+			},
+		},
+		Type: corev1.SecretTypeServiceAccountToken,
+	}
+
+	secret, err = clientset.CoreV1().Secrets(ns).Create(context.Background(), secret, metav1.CreateOptions{})
 	if err != nil {
-		return "", fmt.Errorf("Failed to wait for service account secret: %v", err)
+		return "", fmt.Errorf("Failed to create secret for ServiceAccount %q: %v", sa, err)
 	}
-	token, ok := secret.Data["token"]
-	if !ok {
-		return "", fmt.Errorf("Secret %q for service account %q did not have a token", secret.Name, serviceAccount)
+
+	log.Infof("Created bearer token secret for ServiceAccount %q", sa)
+	serviceAccount.Secrets = append(serviceAccount.Secrets,
+		corev1.ObjectReference{
+			Name:      secret.Name,
+			Namespace: secret.Namespace,
+		})
+
+	_, err = clientset.CoreV1().ServiceAccounts(ns).Update(context.Background(), serviceAccount, metav1.UpdateOptions{})
```

Comment:
Probably needs a timeout.

Also should we use a patch instead of an update, just in case something else about the secret has changed since we fetched it?

Reply:
Changed to patch. Still need to update with timeout.